packages/ra-core/src/controller/input/ReferenceArrayInputBase.tsx

```diff
@@ -0,0 +1,117 @@
+} from './useReferenceArrayInputController';
+import { ResourceContextProvider } from '../../core/ResourceContextProvider';
+import { ChoicesContextProvider } from '../../form/choices/ChoicesContextProvider';
+import { RaRecord } from '../../types';
+import { ChoicesContextValue } from '../../form';
+
+/**
+ * An Input component for fields containing a list of references to another resource.
+ * Useful for 'hasMany' relationship.
+ *
+ * @example
+ * The post object has many tags, so the post resource looks like:
+ * {
+ *    id: 1234,
+ *    tag_ids: [ "1", "23", "4" ]
+ * }
+ *
+ * ReferenceArrayInputBase component fetches the current resources (using
+ * `dataProvider.getMany()`) as well as possible resources (using
+ * `dataProvider.getList()`) in the reference endpoint. It then
+ * delegates rendering to its child component, to which it makes the possible
+ * choices available through the ChoicesContext.
+ *
+ * Use it with a selector component as child, like `<SelectArrayInput>`
+ * or <CheckboxGroupInput>.
+ *
+ * @example
+ * export const PostEdit = () => (
+ *     <Edit>
+ *         <SimpleForm>
+ *             <ReferenceArrayInputBase source="tag_ids" reference="tags">
+ *                 <SelectArrayInput optionText="name" />
+ *             </ReferenceArrayInputBase>
+ *         </SimpleForm>
+ *     </Edit>
+ * );
+ *
+ * By default, restricts the possible values to 25. You can extend this limit
+ * by setting the `perPage` prop.
+ *
+ * @example
+ * <ReferenceArrayInputBase
+ *      source="tag_ids"
+ *      reference="tags"
+ *      perPage={100}>
+ *     <SelectArrayInput optionText="name" />
+ * </ReferenceArrayInputBase>
+ *
+ * By default, orders the possible values by id desc. You can change this order
+ * by setting the `sort` prop (an object with `field` and `order` properties).
+ *
+ * @example
+ * <ReferenceArrayInputBase
+ *      source="tag_ids"
+ *      reference="tags"
+ *      sort={{ field: 'name', order: 'ASC' }}>
+ *     <SelectArrayInput optionText="name" />
+ * </ReferenceArrayInputBase>
+ *
+ * Also, you can filter the query used to populate the possible values. Use the
+ * `filter` prop for that.
+ *
+ * @example
+ * <ReferenceArrayInputBase
+ *      source="tag_ids"
+ *      reference="tags"
+ *      filter={{ is_public: true }}>
+ *     <SelectArrayInput optionText="name" />
+ * </ReferenceArrayInputBase>
+ *
+ * The enclosed component may filter results. ReferenceArrayInputBase create a ChoicesContext which provides
+ * a `setFilters` function. You can call this function to filter the results.
+ */
+export const ReferenceArrayInputBase = <RecordType extends RaRecord = any>(
+    props: ReferenceArrayInputBaseProps<RecordType>
+) => {
+    const { children, filter = defaultFilter, reference, render, sort } = props;
+    if (children && React.Children.count(children) !== 1) {
+        throw new Error(
+            '<ReferenceArrayInputBase> only accepts a single child (like <Datagrid>)'
```

Comment:
Btw I'm just realizing this is wrong, `Datagrid` cant' be a child of `<ReferenceArrayInput>`. We should fix this error in both `ReferenceArrayInputBase` and `ReferenceArrayInput`.